firefox-ios/Client/Frontend/Browser/Tabs/Views/TabDisplayView.swift

```diff
@@ -102,12 +102,15 @@ class TabDisplayView: UIView,
     }
 
     func newState(state: TabsPanelState) {
+        // Temporary workaround to not repeat scroll action for identical state updates
+        guard tabsState != state else { return }
+
         tabsState = state
 
         collectionView.reloadData()
 
-        if let index = state.scrollToIndex {
-            scrollToTab(index)
+        if let scroll = state.scroll {
+            scrollToTab(scroll.toIndex, shouldAnimate: scroll.withAnimation)
         }
```

Comment:
Without this guard scrollToTab could fire again if newState is called whenever the TabsPanelState is changed for an unrelated reason, even if newState was only fired when there is a change.

Reply:
@OrlaM Just to make sure I'm understanding correctly, is there a situation where we mutate `TabsPanelState` outside of its reducer? I've been assuming not.

I agree that we definitely need line 112, but that alone doesn't seem sufficient, we also need the line 106 `guard tabsState != state else { return }` to avoid repeating the same state change -> animation about a dozen times. 🤔 

As I understand it, the reducer should always be initializing the next `TabsPanelState` with `nil` for the `scrollToIndex`/`scroll` property except for an action where we explicitly trigger a scroll (for example, with my new `TabPanelMiddlewareActionType.scrollToTab` action). Thus, `scollToTab()` here would only be triggered once (in theory) when the `TabsPanelState` changes as a result of a `.scrollToTab` action. Then the next tab-related action would create a new `TabsPanelState` update, which would clear the `scrollToIndex`/`scroll` property back to `nil`. No repeated animation.

However, as it stands, the _same_ exact `TabsPanelState` (not a newly initialized one), is being sent multiple times to this `newState()` function because of other global state updates from actions that don't touch `TabsPanelState` (like `FakeReduxActionType`, `MicrosurveyPromptAction`, `PrivateModeActionType`, etc.). So now we get `newState()` called here several times with `state.scroll` as non-nil. 😞  Repeated animations without the line 106 `guard tabsState != state else { return }`.



I actually think it's a very similar situation to this [slack discussion](https://mozilla.slack.com/archives/C05C9RET70F/p1726169897011829?thread_ts=1726169800.683889&cid=C05C9RET70F) where you told Daniel that every time the state is initialized, `showPasswordGenerator` should be explicitly set to false by default. (Now that I'm looking back at his question, I think he must have been seeing exactly what I'm seeing here)

Let me know if there's anything I'm missing here though, I'm still new to our implementation and don't quite have my head wrapped around all the moving pieces. 🙏

Reply:
(This was figured out via a zoom call and has been resolved)